CMakeLists.txt

```diff
@@ -249,7 +260,8 @@ if(BUILD_TESTING)
     target_sources(ninja_test PRIVATE src/includes_normalize_test.cc src/msvc_helper_test.cc
       windows/ninja.manifest)
   endif()
-  target_link_libraries(ninja_test PRIVATE libninja libninja-re2c)
+  find_package(Threads REQUIRED)
+  target_link_libraries(ninja_test PRIVATE libninja libninja-re2c gtest Threads::Threads)
```

Comment:
Thanks for this change.

I found out that this CMakeList.txt file does not work when using the GTEST_ROOT environment variable to point to a local installation (the compiler complains on every `#include <gtest/gtest.h>` line). However, this can be fixed by changing the `gtest` above here into `GTest::gtest` instead.

Reply:
And it seems that using `GTest::gtest` breaks the configure step on GitHub actions (see [1]). Even though it works fine with a locally built 1.10.0 release here. I am so confused :-/ CMake differences maybe?

[1] https://github.com/digit-google/ninja/actions/runs/6786109288

Reply:
Have you tried `GTest::GTest`?

Reply:
It is possible to make all cases work correctly by defining an alias conditionally, i.e.:

```
find_package(Threads REQUIRED)
if (NOT TARGET GTest::gtest)
  add_library(GTest::gtest ALIAS gtest)
endif()
target_link_libraries(ninja_test PRIVATE libninja libninja-re2c GTest::gtest Threads::Threads)
```

Note that `GTest::gtest` is what the official GoogleTest docs are using these days.
I just found this out by trial and error, I don't really understand why CMake's behavior changes here. Any advice welcome :-/

Reply:
Good idea. I just tried `GTest::GTest` locally and it fails in all cases :-(

What is surprising is that even when not using `GTEST_ROOT` (i.e. downloading googletest-1.10.0 and expanding it under build_dir/_deps/googletest-src/), files like `build_dir/_deps/googletest/CMakeFiles/Export/0c08b8e77dd885bfe55a19a9659d9fc1/GTestTargets.cmake` contain definitions for the GTest::gtest target.

However, these are generated when building googletest, but apparently are not available when configuring the build before that. I suspect these are generated by CMake based on input from the googletest build rules.

I am using cmake-3.27.1 by the way.

Reply:
So in other words, when doing a local install of googletest-1.10.0, these CMake files defining GTest::gtest will be copied to the installation directory, then later picked up by find_package() when directed to look there with GTEST_ROOT.

On the other hand, when fetching the package directly from the internet, these files do not exist, and I assume the FetchContent_MakeAvailable() magic can only rely on the top-level library declarations in googletest's own CMakeList.txt files, but does not scope them with `GTest::` (or create an alias like `GTest::gtest` for them).

Reply:
Sounds like a bug in GoogleTest to me. Maybe we could workaround it using `if (TARGET GTest::GTest)`, but I'm not sure if we should.

Out of curiosity: Why are you using GTEST_ROOT in the first place though?

Reply:
Yes, it definitely looks like a GTest bug. I have reproduce the example project from the official GTest documentation at [1], where they use FetchContent_Declare() and FetchContent_MakeAvailable() in the same way as this CMakeLists.txt file, but also links against GTest::gtest.

This works perfectly fine when downloading the archive from the default URL on this page (i.e. https://github.com/google/googletest/archive/03597a01ee50ed33e9dfd640b249b4be3799d395.zip).

However, when I switch it to https://github.com/google/googletest/archive/refs/tags/release-1.10.0.zip, the configure step fails, complaining that there is no GTest::gtest target defined.

Apparently, this has been fixed in 1.11.0 though.

I suggest to conditionally add the ALIAS declaration in the code path that uses FetchContent_Declare() only, and stick to using GTest::gtest as the linked target. This way, updating GoogleTest to a future release will not break the CMake configure step if needed.

[1] https://google.github.io/googletest/quickstart-cmake.html

Reply:
Apart from that, my project (Fuchsia) requires that all prebuilt binaries that we use come from certified and reproducible local CI builds, which must only use sources taken from our own git or archive mirrors, or other "clean" prebuilts generated in the same way. This is both important for legal and technical reasons.

Which means random downloads from the network are not allowed when building these. Hence the need to point to a local installation for dependencies.

Reply:
I have pushed the fix at [1] on my personal github branch, feel free to cherry-pick it onto your branch if you feel so. (My googletest branch is just a rebase of your own on top of current master + my fix, I have no intention to make a PR out of it).

[1] https://github.com/digit-google/ninja/commit/590dd80b2e1c329584ef5789755bdbc0aa2083f3

Reply:
I would guess that the namespace is only applied at install time. My solution for this has been to make a build-time `ALIAS` target of the same name that will be used when installed and then use that everywhere (that [you can](https://gitlab.kitware.com/cmake/cmake/-/issues/19445)).

```cmake
add_library(foo …)
add_library(Foo::foo ALIAS foo)
```